src/middlewared/middlewared/plugins/smb_/util_passdb.py

```diff
@@ -0,0 +1,543 @@
+
+    acct_ctrl, logon_divs, hours_len = unpack('<iHi', entry_bytes[0:10])
+    entry_bytes = entry_bytes[10:]
+
+    hours, entry_bytes = _unpack_samba_pascal_string(entry_bytes, True)
+
+    bad_pw_cnt, logon_cnt = unpack('<HH', entry_bytes[0:4])
+
+    return PDBEntry(
+        username=username,
+        nt_username=nt_username,
+        domain=domain,
+        full_name=full_name,
+        comment=comment,
+        home_dir=homedir,
+        dir_drive=dir_drive,
+        logon_script=logon_script,
+        profile_path=profile_path,
+        acct_desc=acct_desc,
+        acct_ctrl=acct_ctrl,
+        nt_pw=nt_pw.hex().upper(),
+        user_rid=user_rid,
+        group_rid=group_rid,
+        logon_count=logon_cnt,
+        bad_pw_count=bad_pw_cnt,
+        times=times
+    )
+
+
+def _pack_pdb_entry(entry: PDBEntry) -> bytes:
+    """
+    Pack information in PDBEntry into bytes for TDB insertion
+
+    Some values are defaulted to empty strings because we do not
+    provide a mechanism for setting / maintaining them from middleware
+    or explicitly do not support the associated feature (such as lanman password)
+    """
+    data = pack(
+        '<iiiiiii',
+        entry.times.logon,
+        entry.times.logoff,
+        entry.times.kickoff,
+        entry.times.bad_password,
+        entry.times.pass_last_set,
+        entry.times.pass_can_change,
+        entry.times.pass_must_change,
+    )
+
+    data += _pack_samba_pascal_string(entry.username)
+    data += _pack_samba_pascal_string(entry.domain)
+    data += _pack_samba_pascal_string(entry.nt_username)
+    data += _pack_samba_pascal_string(entry.full_name)
+    data += _pack_samba_pascal_string(entry.home_dir)
+    data += _pack_samba_pascal_string(entry.dir_drive)
+    data += _pack_samba_pascal_string(entry.logon_script)
+    data += _pack_samba_pascal_string(entry.profile_path)
+    data += _pack_samba_pascal_string(entry.acct_desc)
+    data += _pack_samba_pascal_string('')  # workstations
+    data += _pack_samba_pascal_string(entry.comment)
+    data += _pack_samba_pascal_string('')  # munged dial
+
+    data += pack('<II', entry.user_rid, entry.group_rid)
+    data += _pack_samba_pascal_string('')  # lanman password
+    data += _pack_samba_pascal_string(bytes.fromhex(entry.nt_pw))
+    data += _pack_samba_pascal_string('')  # NT password history
+    data += pack('<IHi', entry.acct_ctrl, 168, DEFAULT_HOURS_LEN)
+    data += _pack_samba_pascal_string(PACKED_HOURS)
+    data += pack('<HHi', entry.bad_pw_count, entry.logon_count, UNKNOWN_6)
+
+    return data
+
+
+def _parse_passdb_entry(hdl: TDBHandle, tdb_key: str, tdb_val: str) -> PDBEntry:
+    """
+    Retrieve SAMU data based on passdb RID entry and parse bytes into a PDBEntry
+    object.
+
+
+    """
+    try:
+        if (pdb_bytes := hdl.get(f'{USER_PREFIX}{b64decode(tdb_val)[:-1].decode()}')) is None:
+            # malformed passdb entry. Shouldn't happen we'll return None to force
+            # rewrite
```

Comment:
If this shouldn't happen, would it make sense to log so that we'll know that it _**has**_ happened,  (Either here or where ever we handle the `PassdbMustReinit`.)